x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/privileges/ApplicationActionsResolver.java

```diff
@@ -0,0 +1,130 @@
+ * is strictly forbidden unless prior written permission is obtained
+ * from Elasticsearch Incorporated.
+ */
+
+package org.elasticsearch.xpack.idp.privileges;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.security.action.privilege.GetPrivilegesAction;
+import org.elasticsearch.xpack.core.security.action.privilege.GetPrivilegesRequest;
+import org.elasticsearch.xpack.idp.saml.sp.ServiceProviderDefaults;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ApplicationActionsResolver extends AbstractLifecycleComponent {
+
+    private static final int CACHE_SIZE_DEFAULT = 100;
+    private static final TimeValue CACHE_TTL_DEFAULT = TimeValue.timeValueMinutes(90);
+
+    public static final Setting<Integer> CACHE_SIZE
+        = Setting.intSetting("xpack.idp.privileges.cache.size", CACHE_SIZE_DEFAULT, Setting.Property.NodeScope);
+    public static final Setting<TimeValue> CACHE_TTL
+        = Setting.timeSetting("xpack.idp.privileges.cache.ttl", CACHE_TTL_DEFAULT, Setting.Property.NodeScope);
+
+    private final Logger logger = LogManager.getLogger();
+
+    private final ServiceProviderDefaults defaults;
+    private final Client client;
+    private final Cache<String, Set<String>> cache;
+
+    public ApplicationActionsResolver(Settings settings, ServiceProviderDefaults defaults, Client client) {
+        this.defaults = defaults;
+        this.client = new OriginSettingClient(client, ClientHelper.IDP_ORIGIN);
+
+        final TimeValue cacheTtl = CACHE_TTL.get(settings);
+        this.cache = CacheBuilder.<String, Set<String>>builder()
+            .setMaximumWeight(CACHE_SIZE.get(settings))
+            .setExpireAfterWrite(cacheTtl)
+            .build();
+
+        // Preload the cache at 2/3 of its expiry time (TTL). This means that we should never have an empty cache, but if for some reason
+        // the preload thread stops running, we will still automatically refresh the cache on access.
+        final TimeValue preloadInterval = TimeValue.timeValueMillis(cacheTtl.millis() * 2 / 3);
+        client.threadPool().scheduleWithFixedDelay(this::loadPrivilegesForDefaultApplication, preloadInterval, ThreadPool.Names.GENERIC);
```

Comment:
Can you explain why we preload this cache? I think it's because 
- we know we should be getting many requests for the default application but 
- we don't want to statically resolve the actions in here as these are configurable ( even if they probably won't change often ) and
- we don't want to be making GetPrivilege related searches to the security index too often. 

Would this behavior change when we add a [cache for Application Privileges](https://github.com/elastic/elasticsearch/issues/54317) and if so, should we add a TODO or note here ?

Reply:
Realistically we should only have 1 application that we're loading privileges for.
Except that code doesn't currently enforce that from top-to-bottom, so we were left with a weird situation where an "application" string gets passed down when we resolve the privileges. I could have put in a check that `application.equals(defaults.applicationName)`, but decide to just accept that multiple application could be a thing.

Because all (or even most) of the requests are going to be for the same application, when the cache expires we would be likely to have a flood of threads all trying to reload the same data at the same time.
By pre-loading it we can enforce that we have a single request to load the privileges which is going to make a big difference for the load on the security cluster.

But we definitely don't want to cache forever, because adding a new action is a reasonable thing to do, and we don't want to require an IdP restart when it does.

I don't think I would change this with caching on the privileges API - a flood of requests to that action would still be unnecesary and potentially exhaust threadpools etc.